external/fv3fit/fv3fit/emulation/microphysics/layers.py

```diff
@@ -0,0 +1,267 @@
+        if self.selection is not None:
+            tensor = tensor[..., self.selection]
+
+        return tensor
+
+
+class FieldOutput(tf.keras.layers.Layer):
+    """Connect linear dense output layer and denormalize"""
+
+    def __init__(
+        self,
+        sample_out: tf.Tensor,
+        *args,
+        denormalize: Optional[str] = None,
+        enforce_positive=False,
+        **kwargs,
+    ):
+        """
+        Args:
+            sample_out: Output sample for variable to set shape
+                and fit denormalization layer.
+            denormalize: denormalize layer key to use on
+                the dense layer output
+            alt_name: alternative name for unscaled and denorm layers
+        """
+        super().__init__(*args, **kwargs)
+
+        self.unscaled = tf.keras.layers.Dense(
+            sample_out.shape[-1], name=f"unscaled_{self.name}"
+        )
+
+        if denormalize is not None:
+            self.denorm = get_denorm_class(denormalize)(
+                name=f"denormalized_{self.name}"
+            )
+            self.denorm.fit(sample_out)
+        else:
+            self.denorm = tf.keras.layers.Lambda(lambda x: x)
+
+        self.relu = tf.keras.layers.ReLU()
+        self.use_relu = enforce_positive
+
+    def call(self, tensor):
+
+        tensor = self.unscaled(tensor)
+        tensor = self.denorm(tensor)
+
+        if self.use_relu:
+            tensor = self.relu(tensor)
+
+        return tensor
+
+
+class ResidualOutput(tf.keras.layers.Layer):
+    """
+    Add input tensor to an output tensor using timestep increment.
+    Gets net to learn tendency-based updates.
+    """
+
+    def __init__(
+        self,
+        sample_out: tf.Tensor,
+        dt_sec: int,
+        *args,
+        denormalize: Optional[str] = None,
+        enforce_positive: bool = False,
+        **kwargs,
+    ):
+        """
+        Args:
+            sample_out: Output sample for variable to set shape
+                and fit denormalization layer.
+            dt_sec: Timestep length in seconds to use for incrementing
+                input state.
+            denormalize: denormalize layer key to use on
+                the dense layer output
+        """
+        super().__init__(*args, **kwargs)
+
+        self.tendency = FieldOutput(
+            sample_out,
+            denormalize=denormalize,
+            enforce_positive=False,
+            name=f"residual_{self.name}",
+        )
+        self.increment = IncrementStateLayer(dt_sec, name=f"increment_{self.name}")
+        self.use_relu = enforce_positive
+        self.relu = tf.keras.layers.ReLU()
+
+    def call(self, tensors):
+
+        field_input, network_output = tensors
+        tendency = self.tendency(network_output)
+        tensor = self.increment([field_input, tendency])
+
+        if self.use_relu:
+            tensor = self.relu(tensor)
+
+        return tensor
+
+    def get_tendency_output(self, network_output):
```

Comment:
Yes, it could be removed.  I think the current style is a bit more expressive on the dev side since it explicitly calls out what is being grabbed and what the expected input is.  Though, I should probably make the layer attributes "_"-prefixed to express that they are private.